Add concurrency to update command

Coca162 opened the pull request:
Closes #108

This PR is pretty minimal in what it changes, I'd be happy to also work on other features such as adding a max amount of concurrent updates at once or having a flag to revert back to single threaded mode if its deemed desirable. A progress spinner/bar could also be nice to add later.

piegamesde commented:
Thank you for working on this.

I'd prefer if the updating code was transformed into a `future::Stream` pipeline, that way we also get a concurrency limit for free. Bonus points if that structure handles both code paths (update all / update some) without duplication. While draining and rebuilding the `pins` map is not ideal, I consider it acceptable. There *may* be a solution that avoids this, but given that it most likely would involve pinning I wouldn't ask you of trying that unless you already feel at ease with advanced async. Alternatively, `Pin` is `Clone`, simply cloning the data and replacing the old would be an acceptable way of doing this as well for me.

Coca162 commented:
I've changed it up to a `future::Stream` implementation and added the `--conc-count` flag, which someone could probably come up with a better name for.

Coca162 commented:
> There may be a solution that avoids this, but given that it most likely would involve pinning I wouldn't ask you of trying that unless you already feel at ease with advanced async.

I feel like doing this without removing/cloning the elements from the btreemap would require not using tokio's multithreading (since that requires futures to be `'static`) and instead do it using `futures`'s utilities only which is not ideal for larger pin counts. However a benchmark here might be interesting to see if it is noticeably faster at ~1-5 pins to use a singe-threaded concurrent solution instead of a tokio multi-threaded solution.

You could also avoid it by refactoring the entire codebase to switch to reference counted data in `Pin`s but that might not be worth it.

Coca162 commented:
I've updated the PR to use a single-threaded concurrent implementation now with no cloning/removing of pins. I switched to a `pins.pins.iter_mut().filter(|(name, _)| opts.names.is_empty() || opts.names.contains(name))` to get the wanted pins which should be the same for the case where names is empty but for when the user does specify names it will currently ignore nonexistent/duplicate names.

On my PC just by messing about with updating some repositories with 5-8 pins I notice no difference between the performance of the multi-threaded and single-threaded impls.

piegamesde commented on src/cli.rs:
```diff
@@ -656,21 +661,25 @@ impl Opts {
 
-        if opts.names.is_empty() {
-            for (name, pin) in pins.pins.iter_mut() {
-                log::info!("Updating '{}' …", name);
```
This line kinda got lost in the current code revision, wanna add it back in some form or another? Maybe the easiest way it to just put the line into the async block of the update task, however this needs trying out because it might look horrible in practice.

Coca162 replied:
Ideally the way it would be displayed would be using the same sort of idea as [nix-output-monitor](https://github.com/maralorn/nix-output-monitor) where you could see the in-progress updates. I'll add the log back for now but I do find it personally rather annoying. 

Maybe a quick fix would be to only log it if `update_one` takes a significantly long time to update (like it would for nixpkgs)?

Coca162 commented on src/cli.rs:
```diff
@@ -433,6 +437,9 @@ pub struct UpdateOpts {
     pub dry_run: bool,
+    /// Amount of concurrent updates that will be done at once
+    #[structopt(long)]
+    pub conc_count: Option<usize>,
```
Thoughts on `--jobs`? It's a bit generic but people coming from nix would probably understand what it does.

Coca162 commented:
In the most recent commit I added a progress "UI" that much more clearly indicates what is being updated. Let me know what you think!

https://github.com/user-attachments/assets/7aa1a30a-c9fa-4443-ba5b-0f60f6b9c7ef

piegamesde commented:
Oh, this is really neat! Would it be possible to encode the information in color plus some changing text? Also, how does this fall back when not in a terminal?

Coca162 commented:
> Oh, this is really neat! Would it be possible to encode the information in color plus some changing text?

I could add a arg to the pin writer to add a status string as well, maybe it could be turned into a enum later but that might be overkill for now.

> Also, how does this fall back when not in a terminal?

It doesn't use [`is_terminal`](https://doc.rust-lang.org/beta/std/io/trait.IsTerminal.html) so it does not fallback at all! If you want it to do something specific in the case it should let me know.

Coca162 commented:
> Oh, this is really neat! Would it be possible to encode the information in color plus some changing text?

https://github.com/user-attachments/assets/e16d0b04-3571-46f5-a6dd-4d13c4f28caa

piegamesde commented:
> If you want it to do something specific in the case it should let me know.

Just the minimum, as in the software should still be usable without spamming control sequences when its output is piped to a file or a pager for some reason

Coca162 commented:
Nevermind, I also made them just fully ignore the control codes:

https://github.com/user-attachments/assets/9594c70a-7c72-459a-9a1d-8b848f1c0625

piegamesde commented on src/cli.rs:
```diff
@@ -475,6 +482,7 @@ pub enum Command {
     Show,
 
     /// Updates all or the given pin to the latest version.
+    /// Duplicate or missing pins will be ignored.
```
(unresolved)

Coca162 replied:
It's specifically for `--names` to indicate what will happen if you write a name of a pin that doesn't exist. Hopefully it should make more sense now that it is moved to it's parameter-specific documentation.

piegamesde replied:
Ah, now I understand the intent. The issue with the phrasing I see is, that duplicate pins will be ignored in a different way than unspecified ("missing") pins, while referring to the same word in the sentence.

What do you think about "Update only specified pins instead of all"? (and not mention the duplicates because it doesn't add much)

Coca162 replied:
Ah I meant it in a different way then just "unspecified" as well, the code changed the behavior a little to not fail when a user inputs pins that aren't actually in the `sources.json`.

 If I were to rewrite the entire thing ignoring duplicates note it'd be something like:
> Updates only the specified pins.
> Non-existent pins will be ignored.

Coca162 replied:
In the latest commit I just removed this comment and replaced it with some actual warnings for when users do this since otherwise it seems people are going to just get confused even with documentation.

piegamesde commented on src/cli.rs:
```diff
@@ -433,6 +447,9 @@ pub struct UpdateOpts {
     /// Print the diff, but don't write back the changes
     #[structopt(short = "n", long, global = true)]
     pub dry_run: bool,
+    /// Amount of concurrent updates that will be done at once
+    #[structopt(default_value = "5", long)]
+    pub conc_count: usize,
```
```suggestion
    /// Maximum number of concurrent downloads
    #[structopt(default_value = "5", long)]
    pub max_concurrent_downloads: usize,
```

This is the name aria2 uses and I dislike it less than the current proposal, so let's take that instead. The flag may be a bit verbose, but given that most people will never need to change the default this should not be an issue.

piegamesde reviewed (commented):
Getting close, thanks for your work!

The colors in the animation, are they still readable with a light terminal background?

Coca162 commented:
> The colors in the animation, are they still readable with a light terminal background?

https://github.com/user-attachments/assets/723966ee-9f30-4d9f-a695-d9b2de951531

It looks pretty good but ideally the dark/light green would get swapped, though that would probably not be worth the additional code needed for terminal theme detection.

piegamesde reviewed (approved):
This looks awesome! If you want to tidy up the commit history before I merge, do it now

piegamesde commented on src/cli.rs:
```diff
@@ -656,23 +686,67 @@ impl Opts {
+            .pins
+            .iter_mut()
+            .enumerate()
+            .filter(|(_, (name, _))| opts.names.is_empty() || valid_names.contains(name))
```
Shouldn't the enumerate come after the filter for correct indexing?

will commented:
Just updated and got a version of npins with this, and I love it. Thanks!
